Mathlib/Analysis/ODE/MaximalSolution.lean

```diff
@@ -0,0 +1,744 @@
+
+    -- Apply the uniqueness theorem.
+    replace heq_at_t₀ : f₁ t₀ = f₂ t₀ := heq_at_t₀
+    have eq_on_J := ODE_solution_unique_of_mem_Icc_right hv_J
+                      hf₁_cont_J hf₁'_deriv_J hf₁s_univ_J
+                      hf₂_cont_J hf₂'_deriv_J hf₂s_univ_J
+                      heq_at_t₀
+    -- Since t' ∈ J, f₁(t') = f₂(t').
+    exact eq_on_J (right_mem_Icc.mpr h_t₀_le_t')
+
+  · -- Case 2: t' < t₀. Apply uniqueness on the closed interval J = [t', t₀].
+    let J := Icc t' t₀
+    -- J is contained in K_int.
+    have hJ_sub_K_int : J ⊆ K_int := by
+      intro j hj_in_J
+      exact hK_int_conn.isPreconnected.ordConnected.out ht'_in_K_int ht₀_mem_K_int hj_in_J
+
+    -- The Lipschitz condition on v holds for t ∈ Ioc t' t₀.
+    have hv_J : ∀ (t_val : ℝ) (ht_val : t_val ∈ Ioc t' t₀),
+        LipschitzOnWith K_const (v t_val) univ := by
+      intro t_val ht_val_in_Ioc
+      exact (h_lipschitz t_val (hJ_sub_K_int (mem_Icc_of_Ioc ht_val_in_Ioc))).lipschitzOnWith
+
+    -- Verify conditions for ODE_solution_unique_of_mem_Icc_left for f₁.
+    have hf₁_cont_J : ContinuousOn f₁ J :=
+      h₁.continuousOn.mono (Subset.trans hJ_sub_K_int inter_subset_left)
+    have hf₁'_deriv_J : ∀ (t_val : ℝ) (ht_val : t_val ∈ Ioc t' t₀),
+        HasDerivWithinAt f₁ (v t_val (f₁ t_val)) (Iic t_val) t_val := by
+      intro t_val ht_val_in_Ioc
+      have deriv_at := h₁.deriv_eq t_val
+        (Subset.trans hJ_sub_K_int inter_subset_left (mem_Icc_of_Ioc ht_val_in_Ioc))
+      exact deriv_at.hasDerivWithinAt
+    have hf₁s_univ_J :
+      ∀ (t_val : ℝ) (ht_val : t_val ∈ Ioc t' t₀), f₁ t_val ∈ univ := fun _ _ => trivial
+
+    -- Verify conditions for ODE_solution_unique_of_mem_Icc_left for f₂.
+    have hf₂_cont_J : ContinuousOn f₂ J :=
+      h₂.continuousOn.mono (Subset.trans hJ_sub_K_int inter_subset_right)
+    have hf₂'_deriv_J : ∀ (t_val : ℝ) (ht_val : t_val ∈ Ioc t' t₀),
+        HasDerivWithinAt f₂ (v t_val (f₂ t_val)) (Iic t_val) t_val := by
+      intro t_val ht_val_in_Ioc
+      have deriv_at := h₂.deriv_eq t_val
+        (Subset.trans hJ_sub_K_int inter_subset_right (mem_Icc_of_Ioc ht_val_in_Ioc))
+      exact deriv_at.hasDerivWithinAt
+    have hf₂s_univ_J :
+      ∀ (t_val : ℝ) (ht_val : t_val ∈ Ioc t' t₀), f₂ t_val ∈ univ := fun _ _ => trivial
+
+    -- Apply the uniqueness theorem (solutions agree at the right endpoint t₀).
+    replace heq_at_t₀ : f₁ t₀ = f₂ t₀ := heq_at_t₀
+    have eq_on_J := ODE_solution_unique_of_mem_Icc_left hv_J
+                      hf₁_cont_J hf₁'_deriv_J hf₁s_univ_J
+                      hf₂_cont_J hf₂'_deriv_J hf₂s_univ_J
+                      heq_at_t₀
+    -- Since t' ∈ J, f₁(t') = f₂(t').
+    exact eq_on_J (left_mem_Icc.mpr h_t'_le_t₀)
+
+/--
+A solution `(f, I)` to the ODE `x' = v(t, x)` with initial value `(t₀, x₀)` is maximal if it
+cannot be extended to a solution on any strictly larger open connected domain `J`.
+-/
+structure IsMaximalODESolution (f : ℝ → E) (I : Set ℝ) : Prop
```

Comment:
Again this doesn't need `t₀ x₀` in the definition, but with the removal of assumptions on `I` from `IsODESolution`, you may want to add those assumptions here.